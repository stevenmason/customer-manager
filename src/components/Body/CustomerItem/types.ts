import { Customer } from "../../../store/types";
import { setEditingCustomer, deleteCustomer } from "../../../store/actions";

export interface Props {
  customer: Customer;
  setEditingCustomer: typeof setEditingCustomer;
  deleteCustomer: typeof deleteCustomer;
}
