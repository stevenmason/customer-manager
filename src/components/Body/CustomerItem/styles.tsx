import styled from "styled-components";

const Text = styled.span`
  display: block;
`;

const Container = styled.span`
  display: flex;
`;

export { Text, Container };
