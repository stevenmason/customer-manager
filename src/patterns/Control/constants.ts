enum SIZES {
  LARGE = "LARGE",
  SMALL = "SMALL"
}

export { SIZES };
