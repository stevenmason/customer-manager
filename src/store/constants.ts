enum ACTIONS {
  SET_CUSTOMERS = "SET_CUSTOMERS",
  SET_CUSTOMER = "SET_CUSTOMER",
  DELETE_CUSTOMER = "DELETE_CUSTOMER",
  SET_EDITING_CUSTOMER = "SET_EDITING_CUSTOMER",
  SET_EDITING_CUSTOMER_VALUE = "SET_EDITING_CUSTOMER_VALUE",
  SET_LOADING = "SET_LOADING",
  SET_FILTER = "SET_FILTER"
}

const DEFAULT_CUSTOMER = {
  id: "-1",
  firstName: "",
  lastName: "",
  dateOfBirth: ""
};

export { ACTIONS, DEFAULT_CUSTOMER };
