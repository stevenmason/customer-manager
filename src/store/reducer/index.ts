import { combineReducers } from "redux";

import customers from "./customers";
import editingCustomer from "./editingCustomer";
import filter from "./filter";
import loading from "./loading";

const rootReducer = combineReducers({
  customers,
  filter,
  editingCustomer,
  loading
});

export default rootReducer;
