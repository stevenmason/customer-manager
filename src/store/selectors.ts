import { createSelector } from "reselect";

import { State, Customer } from "./types";
import { customerContainsFilter } from "./utils";

const customers = (state: State) => state.customers;
const loading = (state: State) => state.loading;
const filter = (state: State) => state.filter;
const editingCustomer = (state: State) => state.editingCustomer;

const filteredCustomers = createSelector(
  [customers, filter],
  (customers, filter) =>
    Object.values(customers).filter((customer: Customer) =>
      customerContainsFilter(customer, filter)
    )
);

export { filteredCustomers, loading, filter, editingCustomer };
