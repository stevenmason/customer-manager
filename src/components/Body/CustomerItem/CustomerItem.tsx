import React from "react";

import * as SC from "./styles";
import { Props } from "./types";
import Item from "../../../patterns/List/Item";
import { EditIcon, DeleteIcon } from "../../../icons";
import Control, { CONTROL_SIZES } from "../../../patterns/Control";

const CustomerItem: React.FC<Props> = ({
  customer,
  setEditingCustomer,
  deleteCustomer
}) => {
  const { firstName, dateOfBirth, id, lastName } = customer;
  return (
    <Item key={id}>
      <div>
        <Control size={CONTROL_SIZES.SMALL} label="Name:">
          <SC.Text>{`${firstName} ${lastName}`}</SC.Text>
        </Control>
        <Control size={CONTROL_SIZES.SMALL} label="Date of Birth:">
          <SC.Text>{new Date(dateOfBirth).toDateString()}</SC.Text>
        </Control>
      </div>
      <SC.Container>
        <EditIcon onClick={() => setEditingCustomer(customer)} />
        <DeleteIcon onClick={() => deleteCustomer(customer.id)} />
      </SC.Container>
    </Item>
  );
};

export default CustomerItem;
