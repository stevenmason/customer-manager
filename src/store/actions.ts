import { uniqueId } from "lodash";

import { ACTIONS, DEFAULT_CUSTOMER } from "./constants";
import { Customer } from "./types";

const setCustomers = (customers: Customer[]) => ({
  type: ACTIONS.SET_CUSTOMERS as ACTIONS.SET_CUSTOMERS,
  customers
});

const setCustomer = (customer: Customer) => ({
  type: ACTIONS.SET_CUSTOMER as ACTIONS.SET_CUSTOMER,
  customer
});

const deleteCustomer = (id: string) => ({
  type: ACTIONS.DELETE_CUSTOMER as ACTIONS.DELETE_CUSTOMER,
  id
});

const setEditingCustomer = (editingCustomer: Customer | null) => ({
  type: ACTIONS.SET_EDITING_CUSTOMER as ACTIONS.SET_EDITING_CUSTOMER,
  editingCustomer
});

const addNewEditingCustomer = () => ({
  type: ACTIONS.SET_EDITING_CUSTOMER as ACTIONS.SET_EDITING_CUSTOMER,
  editingCustomer: {
    ...DEFAULT_CUSTOMER,
    id: uniqueId()
  }
});

const setEditingCustomerValue = (value: string, name: string) => ({
  type: ACTIONS.SET_EDITING_CUSTOMER_VALUE as ACTIONS.SET_EDITING_CUSTOMER_VALUE,
  name,
  value
});

const setLoading = (loading: boolean) => ({
  type: ACTIONS.SET_LOADING as ACTIONS.SET_LOADING,
  loading
});

const setFilter = (filter: string) => ({
  type: ACTIONS.SET_FILTER as ACTIONS.SET_FILTER,
  filter
});

export {
  setCustomers,
  setCustomer,
  setEditingCustomer,
  setEditingCustomerValue,
  setLoading,
  setFilter,
  addNewEditingCustomer,
  deleteCustomer
};
