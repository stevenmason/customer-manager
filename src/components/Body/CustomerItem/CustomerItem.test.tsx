import React from "react";
import renderer from "react-test-renderer";

import CustomerItem from "./CustomerItem";
import { DEFAULT_CUSTOMER } from "../../../store/constants";
import { setEditingCustomer, deleteCustomer } from "../../../store/actions";

const PROPS = {
  customer: DEFAULT_CUSTOMER,
  setEditingCustomer,
  deleteCustomer
};

it("renders correctly", () => {
  const tree = renderer.create(<CustomerItem {...PROPS} />);
  expect(tree.toJSON()).toMatchSnapshot();
});
