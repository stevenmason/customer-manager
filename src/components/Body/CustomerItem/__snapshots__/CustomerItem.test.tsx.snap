// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-htpNat dISaAK"
>
  <div>
    <div
      className="sc-ifAKCX LcqdU"
      size="SMALL"
    >
      <label
        className="sc-EHOje hxhCYX"
      >
        Name:
      </label>
      <span
        className="sc-bdVaJa fpcbkF"
      >
         
      </span>
    </div>
    <div
      className="sc-ifAKCX LcqdU"
      size="SMALL"
    >
      <label
        className="sc-EHOje hxhCYX"
      >
        Date of Birth:
      </label>
      <span
        className="sc-bdVaJa fpcbkF"
      >
        Invalid Date
      </span>
    </div>
  </div>
  <span
    className="sc-bwzfXH jIEWCi"
  >
    <div
      className="sc-bxivhb isqwrT"
      onClick={[Function]}
    >
      <svg
        viewBox="0 0 24 24"
      >
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </div>
    <div
      className="sc-bxivhb isqwrT"
      onClick={[Function]}
    >
      <svg
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </div>
  </span>
</div>
`;
